src/artifacts.js

```diff
@@ -423,6 +643,119 @@ var replyWithArtifact = async function(taskId, runId, name, req, res) {
+  route:      '/task/:taskId/runs/:runId/artifacts/:name(*)',
+  name:       'completeArtifact',
+  stability:  API.stability.experimental,
+  scopes: [
+    [
+      'queue:create-artifact:<name>',
+      'assume:worker-id:<workerGroup>/<workerId>',
+    ], [
+      'queue:create-artifact:<taskId>/<runId>',
+    ],
+  ],
+  deferAuth:  true,
+  input:      'put-artifact-request.json#',
+  output:     'put-artifact-response.json#',
+  title:      'Complete Artifact',
+  description: 'tbd',
+}, async function(req, res) {
+  let taskId      = req.params.taskId;
+  let runId       = parseInt(req.params.runId, 10);
+  let name        = req.params.name;
+  let input       = req.body;
+  let isPublic    = /^public\//.test(name);
+
+  // Hmm, list destructuring is probably better here
+  let [artifact, task] = await Promise.all([
+    this.Artifact.load({taskId, runId, name}, true),
+    this.Task.load({taskId}, true),
+  ]);
+
+  // TODO: Have a shortcut here where things which are already
+  // marked as present short circuit
+
+  let run = task.runs[runId];
+  if (!run) {
+    return res.reportError('InputError',
+      'Run not found',
+      {});
+  }
+  let workerGroup = run.workerGroup;
+  let workerId = run.workerId;
+
+  // Authenticate request by providing parameters
+  if (!req.satisfies({
+    taskId,
+    runId,
+    workerGroup,
+    workerId,
+    name,
+  })) {
+    return;
+  }
+
+  if (artifact.storageType === 'blob') {
+    let etag;
+    if (artifact.details.partsHash) {
+      etag = await this.s3Controller.completeMultipartUpload({
+        bucket: artifact.details.bucket,
+        key: artifact.details.key,
+        etags: input.etags,
+        uploadId: artifact.details.uploadId,
+        tags: {taskId, runId, name},
+      });
+    } else {
+      let url = await this.s3Controller.generateGetUrl({
+        bucket: artifact.details.bucket,
+        key: artifact.details.key,
+        signed: !isPublic,
+      });
+
+      if (urllib.parse(url).protocol !== 'https:') {
+        throw new Error('Only HTTPS is allowed for artifacts');
+      }
+
+      let headRes = await this.s3Runner.run({
+        req:{
+          url,
+          method: 'HEAD',
+          headers: {},
+        },
+      });
+      
+      if (headRes.headers['x-amz-meta-content-sha256'] !== artifact.details.contentSha256) {
+        throw new Error('S3 object does not have correct Content-Sha256 metadata');
+      }
+      
+      etag = input.etags[0];
+    }
+
+    await artifact.modify((artifact) => {
+      artifact.details.etag = etag;
+      // Now that we're finished, we don't want to store the uploadId any longer
+      artifact.details = _.omit(artifact.details, 'uploadId');
+      artifact.present = true;
+    });
+
+    await this.publisher.artifactCreated({
+      status: task.status(),
+      artifact: artifact.json(),
+      workerGroup,
+      workerId,
+      runId,
+    }, task.routes);
+    return res.reply({
+      outcome: 'success',
```

Comment:
Just do: `return res.status(200).json({})`, we don't have any content here...

Reply:
Or maybe do: `return res.reply({artifact: artifact.json()})` that's a somewhat useful thing to respond and consistent with what we return in listing of artifacts :)